tasks/libs/pipeline/failure_summary.py

```diff
@@ -0,0 +1,331 @@
+    listing = json.loads(listing)
+    listing = [item['Key'].removeprefix(FAILURE_SUMMARY_S3_PREFIX + '/') for item in listing['Contents']]
+    listing = [item for item in listing if item.endswith('.json')]
+
+    return listing
+
+
+def is_valid_job(repo: Project, job: ProjectPipelineJob) -> bool:
+    """
+    Returns whether the job is finished (failed / success) and if it is not an infrastructure failure
+    """
+    # Not finished
+    if job.status not in ['failed', 'success']:
+        return False
+
+    # Ignore infra failures
+    if job.status == 'failed':
+        trace = str(repo.jobs.get(job.id, lazy=True).trace(), 'utf-8')
+        failure_type = get_infra_failure_info(trace)
+        if failure_type is not None:
+            return False
+
+    return True
+
+
+def fetch_jobs(ctx: Context, pipeline_id: int) -> SummaryData:
+    """
+    Returns all the jobs for a given pipeline
+    """
+    id = int(datetime.now(timezone.utc).timestamp())
+    repo = get_gitlab_repo()
+
+    jobs: list[ProjectPipelineJob] = []
+    pipeline = repo.pipelines.get(pipeline_id, lazy=True)
+    for job in pipeline.jobs.list(per_page=100, all=True):
+        if is_valid_job(repo, job):
+            jobs.append(job)
+
+    return SummaryData(ctx=ctx, id=id, jobs=jobs, pipeline=pipeline)
+
+
+def fetch_summaries(ctx: Context, period: timedelta) -> SummaryData:
+    """
+    Returns all summaries for a given period
+    """
+    ids = SummaryData.list_summaries(ctx, after=int((datetime.now(timezone.utc) - period).timestamp()))
+    repo = get_gitlab_repo()
+    summaries = [SummaryData.read(ctx, repo, id) for id in ids]
+    summary = SummaryData.merge(summaries)
+
+    return summary
+
+
+def upload_summary(ctx: Context, pipeline_id: int) -> SummaryData:
+    """
+    Creates and uploads a summary for a given pipeline
+    """
+    summary = fetch_jobs(ctx, pipeline_id)
+    summary.write()
+
+    return summary
+
+
+def send_summary_slack_notification(channel: str, stats: list[dict], allow_failure: bool = False):
+    """
+    Send the summary to channel with these job stats
+    - stats: Item of the dict returned by SummaryStats.make_stats
+    """
+    # Avoid circular dependency
+    from slack_sdk import WebClient
+
+    from tasks.notify import NOTIFICATION_DISCLAIMER, get_ci_visibility_job_url
+
+    # Do not send notification
+    if not stats:
+        return
+
+    # Create message
+    not_allowed_query = '-' if not allow_failure else ''
+    period = 'Daily' if not allow_failure else 'Weekly'
+    duration = '24 hours' if not allow_failure else 'week'
+    delta = timedelta(days=1) if not allow_failure else timedelta(weeks=1)
+    you_own = ' you own' if channel != GITHUB_SLACK_MAP[ALL_TEAMS] else ''
+    flaky_tests = (
+        ''
+        if allow_failure
+        else ' In case of tests, you can <https://datadoghq.atlassian.net/wiki/spaces/ADX/pages/3405611398/Flaky+tests+in+go+introducing+flake.Mark|mark them as flaky>.'
```

Comment:
```suggestion
        else ' In case of flaky tests, you can <https://datadoghq.atlassian.net/wiki/spaces/ADX/pages/3405611398/Flaky+tests+in+go+introducing+flake.Mark|mark them as flaky>.'
```